File: IGListTest/IGListTest/SectionControllers/JournalSectionController.m
//
//  JournalSectionController.m
//  IGListTest
//
//  Created by liuruixuan on 2023/4/16.
//

#import "JournalSectionController.h"
#import "JournalEntryCell.h"
#import "MarslinkModel.h"
#import "JournalEntryDateCell.h"

@interface JournalSectionController ()

@end

@implementation JournalSectionController

- (instancetype)init
{
    self = [super init];
    if (self) {
        self.inset = UIEdgeInsetsMake(0, 0, 15, 0);
    }
    return self;
}

- (NSInteger)numberOfItems {
    return 2;
}

- (CGSize)sizeForItemAtIndex:(NSInteger)index {
    if (index == 0) {
        return CGSizeMake(self.collectionContext.containerSize.width, 30);
    } else {
        return [JournalEntryCell cellSizeWithWidth:self.collectionContext.containerSize.width text:self.entry.text];
    }
}

- (__kindof UICollectionViewCell *)cellForItemAtIndex:(NSInteger)index {
    
    if (index == 0) {
        JournalEntryDateCell *cell = [self.collectionContext dequeueReusableCellOfClass:[JournalEntryDateCell class] forSectionController:self atIndex:index];
        cell.label.text = [NSString stringWithFormat:@"SOL %ld", (long)[self.solFormatter solsFromDate:self.entry.date]];
        return cell;
    } else {
        JournalEntryCell *cell = [self.collectionContext dequeueReusableCellOfClass:[JournalEntryCell class] forSectionController:self atIndex:index];
        cell.label.text = self.entry.text;
        return cell;
    }
    
}

- (void)didUpdateToObject:(id)object {
    self.entry = (JournalEntry *)object;
}

- (void)didSelectItemAtIndex:(NSInteger)index {
    
}

- (SolFormatter *)solFormatter {
    if (!_solFormatter) {
        _solFormatter = [[SolFormatter alloc] init];
    }
    return _solFormatter;
}

@end

File: Marslink_Swift5/Marslink_Swift5/SectionControllers/JournalSectionController.swift
//
//  JournalSectionController.swift
//  Marslink_Swift5
//
//  Created by liuruixuan on 2023/4/17.
//

import IGListKit
import UIKit

class JournalSectionController: ListSectionController {
    var entry: JournalEntry!
    let solFormatter = SolFormatter()

    override init() {
        super.init()
        inset = UIEdgeInsets(top: 0, left: 0, bottom: 15, right: 0)
    }

    override func numberOfItems() -> Int {
        return 2
    }

    override func sizeForItem(at index: Int) -> CGSize {
        guard let context = collectionContext, let entry = entry else { return .zero }
        let width = context.containerSize.width

        if index == 0 {
            return CGSize(width: width, height: 30)
        } else {
            return JournalEntryCell.cellSize(width: width, text: entry.text)
        }
    }

    override func cellForItem(at index: Int) -> UICollectionViewCell {
        let cellClass: AnyClass = index == 0 ? JournalEntryDateCell.self : JournalEntryCell.self
        let cell = collectionContext!.dequeueReusableCell(of: cellClass, for: self, at: index)

        if let cell = cell as? JournalEntryDateCell {
            cell.label.text = "SOL \(solFormatter.sols(fromDate: entry.date))"
        } else if let cell = cell as? JournalEntryCell {
            cell.label.text = entry.text
        }
        return cell
    }

    override func didUpdate(to object: Any) {
        entry = object as? JournalEntry
    }

    override func didSelectItem(at index: Int) {
    }
}
